File: backend/src/domain/valuesobject/Price.go
package valuesobject

import "errors"

type Price struct {
	price float64
}

func NewPrice(price float64) (*Price, error) {
	if err := validatePrice(price); err != nil {
		return nil, err
	}
	return &Price{price: price}, nil
}

func validatePrice(price float64) error {
	if  price <= 0 {
		return errors.New("preço não pode ser 0 ou menor do que zero")
	}
	return nil
}

func (p *Price) GetValue() float64 {
	return p.price
}

File: frontend/src/domain/valuesobject/Price.ts
export class Price {
  private constructor(private price: number) {}

  public static create(price: number): Price {
    this.validate(price);
    return new Price(price);
  }
  private static validate(price: number): void {
    if (!price || typeof price !== 'number' || price <= 0) {
      throw new Error('preço é obrigatorio e não poder ser menor que zero');
    }
  }
  public getValue(): number {
    return this.price;
  }
}
